### PR TITLE
Periodic flush on in_tail plugin for multiline format #668

### PR DESCRIPTION
**Pull request by kikumoto**
I wrote patches for #668.
Please check this.
Thanks.

### CODE REVIEW

**Comment by repeatedly** [comment]
We need testers for this patch.

File: lib/fluent/plugin/in_tail.rb [hunk 1]
```diff
@@ -262,6 +264,7 @@ def parse_multilines(lines, tail_watcher)
       lb = tail_watcher.line_buffer
       es = MultiEventStream.new
       if @parser.has_firstline?
+        tail_watcher.line_buffer_flusher.reset unless tail_watcher.line_buffer_flusher.nil?
```
Comment (sonots) [inline lib/fluent/plugin/in_tail.rb]: `if tail_watcher.line_buffer_flusher`

File: lib/fluent/plugin/in_tail.rb [hunk 2]
```diff
@@ -580,6 +586,29 @@ def on_notify
...
+          if !@start.nil? && !@flush_interval.nil?
+            if Time.now - @start >= @flush_interval
+              @flush_method.call(tw)
+              tw.line_buffer = nil
+              @start = nil
+            end
+          end
+        end
+
+        def reset
```
Comment (sonots) [inline lib/fluent/plugin/in_tail.rb]: `reset_timer` or something looks more descriptive.

File: lib/fluent/plugin/in_tail.rb [hunk 3]
```diff
@@ -580,6 +586,29 @@ def on_notify
           @log.error_backtrace
         end
       end
+
+      class LineBufferFlusher
```
Comment (sonots) [inline lib/fluent/plugin/in_tail.rb]: `LineBufferIntervalFlusher` or `LineBufferTimerFlusher` looks more descriptive. If we say `LineBufferFlusher`, it must be responsible for all kinds of line buffer flushing, but this class is not.

File: lib/fluent/plugin/in_tail.rb [hunk 4]
```diff
@@ -580,6 +586,29 @@ def on_notify
...
+              tw.line_buffer = nil
+              @start = nil
+            end
+          end
+        end
+
+        def reset
+          @start = Time.now
+        end
+       end
```
Comment (sonots) [inline lib/fluent/plugin/in_tail.rb]: indent

File: lib/fluent/plugin/in_tail.rb [hunk 5]
```diff
@@ -580,6 +586,29 @@ def on_notify
...
+      class LineBufferFlusher
+        def initialize(log, flush_interval, &flush_method)
+          @log = log
+          @flush_interval = flush_interval
+          @flush_method = flush_method
+          @start = nil
+        end
+
+        def on_notify(tw)
+          if !@start.nil? && !@flush_interval.nil?
```
Comment (sonots) [inline lib/fluent/plugin/in_tail.rb]: `if @start && @flush_interval`

**Comment by kikumoto** [comment]
I corrected the points that were pointed out.
But TravisCI have been failed. 
It seems that it failed at the place where this PR was not related to.
What I should do ?
Please check. 
thank you.

**Comment by sonots** [comment]
I re-ran tests

**Comment by kikumoto** [comment]
Thanks!!!

**Comment by repeatedly** [comment]
Ah, wait. This patch is for v0.12.
Could you send the patch to master?

**Comment by repeatedly** [comment]
After update patch and send to master, I will merge it.

**Comment by kikumoto** [comment]
I send the patch to master with https://github.com/fluent/fluentd/pull/775 .
thanks.

**Comment by repeatedly** [comment]
#775 is merged. Thanks!

**Pull request closed without merging**